File: src/main/java/fr/vlegall/sochief/service/recipe/RecipeRefService.java
package fr.vlegall.sochief.service.recipe;

import fr.vlegall.sochief.contracts.common.IdOrNameDto;
import fr.vlegall.sochief.exception.NotFoundException;
import fr.vlegall.sochief.model.recipe.*;
import fr.vlegall.sochief.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class RecipeRefService {
    private final RecipeCategoryRepository categoryRepo;
    private final RecipeDifficultyRepository difficultyRepo;
    private final IngredientRepository ingredientRepo;
    private final IngredientUnitRepository unitRepo;
    private final TagRepository tagRepo;
    private final UtensilRepository utensilRepo;

    public RecipeRefService(RecipeCategoryRepository categoryRepo,
                            RecipeDifficultyRepository difficultyRepo,
                            IngredientRepository ingredientRepo,
                            IngredientUnitRepository unitRepo,
                            TagRepository tagRepo,
                            UtensilRepository utensilRepo) {
        this.categoryRepo = categoryRepo;
        this.difficultyRepo = difficultyRepo;
        this.ingredientRepo = ingredientRepo;
        this.unitRepo = unitRepo;
        this.tagRepo = tagRepo;
        this.utensilRepo = utensilRepo;
    }

    @Transactional(readOnly = true)
    public RecipeCategory getCategory(Long categoryId) {
        return categoryRepo.findById(categoryId)
                .orElseThrow(() -> new NotFoundException("RecipeCategory " + categoryId + " not found"));
    }

    @Transactional(readOnly = true)
    public RecipeDifficulty getDifficulty(Long difficultyId) {
        return difficultyRepo.findById(difficultyId)
                .orElseThrow(() -> new NotFoundException("RecipeDifficulty " + difficultyId + " not found"));
    }

    @Transactional
    public Ingredient resolveIngredient(IdOrNameDto ref) {
        return resolveEntity(
                ref,
                id -> ingredientRepo.findById(id).orElse(null),
                ingredientRepo::findByNameIgnoreCase,
                name -> ingredientRepo.save(new Ingredient(null, name))
                , "Ingredient");
    }

    @Transactional
    public IngredientUnit resolveUnit(IdOrNameDto ref) {
        return resolveEntity(
                ref,
                id -> unitRepo.findById(id).orElse(null),
                unitRepo::findByNameIgnoreCase,
                name -> unitRepo.save(new IngredientUnit(null, name))
                , "IngredientUnit");
    }

    @Transactional
    public Tag resolveTag(IdOrNameDto ref) {
        return resolveEntity(
                ref,
                id -> tagRepo.findById(id).orElse(null),
                tagRepo::findByNameIgnoreCase,
                name -> tagRepo.save(new Tag(null, name))
                , "Tag");
    }

    @Transactional
    public Utensil resolveUtensil(IdOrNameDto ref) {
        return resolveEntity(
                ref,
                id -> utensilRepo.findById(id).orElse(null),
                utensilRepo::findByNameIgnoreCase,
                name -> utensilRepo.save(new Utensil(null, name))
                , "Utensil");
    }

    private <T> T resolveEntity(IdOrNameDto ref,
                                java.util.function.Function<Long, T> byId,
                                java.util.function.Function<String, T> byName,
                                java.util.function.Function<String, T> create,
                                String label) {
        if (ref.getId() != null) {
            T found = byId.apply(ref.getId());
            if (found == null) throw new NotFoundException(label + " " + ref.getId() + " not found");
            return found;
        }
        String name = ref.getName() != null ? ref.getName().trim() : "";
        if (name.isBlank()) {
            throw new IllegalArgumentException(label + " reference must contain id or name");
        }
        T found = byName.apply(name);
        return found != null ? found : create.apply(name);
    }
}

File: src/main/kotlin/fr/vlegall/sochief/service/recipe/RecipeRefService.kt
package fr.vlegall.sochief.service.recipe

import fr.vlegall.sochief.exception.NotFoundException
import fr.vlegall.sochief.model.recipe.*
import fr.vlegall.sochief.repository.*
import fr.vlegall.sochief.contracts.common.IdOrNameDto
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class RecipeRefService(
    private val categoryRepo: RecipeCategoryRepository,
    private val difficultyRepo: RecipeDifficultyRepository,
    private val ingredientRepo: IngredientRepository,
    private val unitRepo: IngredientUnitRepository,
    private val tagRepo: TagRepository,
    private val utensilRepo: UtensilRepository,
) {

    @Transactional(readOnly = true)
    fun getCategory(categoryId: Long): RecipeCategory =
        categoryRepo.findById(categoryId).orElseThrow {
            NotFoundException("RecipeCategory $categoryId not found")
        }

    @Transactional(readOnly = true)
    fun getDifficulty(difficultyId: Long): RecipeDifficulty =
        difficultyRepo.findById(difficultyId).orElseThrow {
            NotFoundException("RecipeDifficulty $difficultyId not found")
        }

    @Transactional
    fun resolveIngredient(ref: IdOrNameDto): Ingredient =
        resolveEntity(
            ref = ref,
            byId = { ingredientRepo.findById(it).orElse(null) },
            byName = { ingredientRepo.findByNameIgnoreCase(it) },
            create = { ingredientRepo.save(Ingredient(name = it)) },
            label = "Ingredient"
        )

    @Transactional
    fun resolveUnit(ref: IdOrNameDto): IngredientUnit =
        resolveEntity(
            ref = ref,
            byId = { unitRepo.findById(it).orElse(null) },
            byName = { unitRepo.findByNameIgnoreCase(it) },
            create = { unitRepo.save(IngredientUnit(name = it)) },
            label = "IngredientUnit"
        )

    @Transactional
    fun resolveTag(ref: IdOrNameDto): Tag =
        resolveEntity(
            ref = ref,
            byId = { tagRepo.findById(it).orElse(null) },
            byName = { tagRepo.findByNameIgnoreCase(it) },
            create = { tagRepo.save(Tag(name = it)) },
            label = "Tag"
        )

    @Transactional
    fun resolveUtensil(ref: IdOrNameDto): Utensil =
        resolveEntity(
            ref = ref,
            byId = { utensilRepo.findById(it).orElse(null) },
            byName = { utensilRepo.findByNameIgnoreCase(it) },
            create = { utensilRepo.save(Utensil(name = it)) },
            label = "Utensil"
        )

    private fun <T> resolveEntity(
        ref: IdOrNameDto,
        byId: (Long) -> T?,
        byName: (String) -> T?,
        create: (String) -> T,
        label: String
    ): T {
        ref.id?.let { id ->
            return byId(id) ?: throw NotFoundException("$label $id not found")
        }
        val name = ref.name?.trim().orEmpty()
        if (name.isBlank()) throw IllegalArgumentException("$label reference must contain id or name")
        return byName(name) ?: create(name)
    }
}